src/main/scala/edu/ie3/simona/agent/grid/congestion/VoltageRange.scala

```diff
@@ -0,0 +1,318 @@
+      deltaMinus: ComparableQuantity[Dimensionless],
+  ): VoltageRange = {
+
+    val plus = deltaPlus.getValue.doubleValue()
+    val minus = deltaMinus.getValue.doubleValue()
+
+    val value = if plus > minus then {
+      // we could have a voltage violation of one limit
+      (plus + minus) / 2
+    } else {
+      (plus > 0, minus < 0) match {
+        case (true, false) =>
+          // we have a voltage violation of the lower limit
+          // since the upper limit is fine, we can increase the voltage a bit
+          plus
+        case (false, true) =>
+          // we have a voltage violation of the upper limit
+          // since the lower limit is fine, we can decrease the voltage a bit
+          minus
+        case _ =>
+          // we have a voltage violation of both limits, we can't fix this
+          0
+      }
+    }
+
+    val factor = 1e3
+
+    val suggestion = if value < 0 then {
+      (value * factor).floor / factor
+    } else {
+      (value * factor).ceil / factor
+    }
+
+    // check if tapping is required
+    if plus < 0 || minus > 0 then {
+      VoltageRange(
+        deltaPlus,
+        deltaMinus,
+        suggestion.asPu,
+      )
+    } else {
+      // the voltage in this range is fine, set the suggested voltage change to zero
+      VoltageRange(
+        deltaPlus,
+        deltaMinus,
+        0.asPu,
+      )
+    }
+  }
+
+  /** Method to calculate the possible range of voltage changes.
+    *
+    * @param powerFlowResultEvent
+    *   Results from simulating the grid.
+    * @param voltageLimits
+    *   Voltage limits.
+    * @param gridComponents
+    *   All components of the grid.
+    * @param inferiorData
+    *   Map: inferior grid to [[VoltageRange]] and [[TransformerTappingModel]].
+    * @return
+    *   A [[VoltageRange]].
+    */
+  def apply(
+      powerFlowResultEvent: PowerFlowResultEvent,
+      voltageLimits: VoltageLimits,
+      gridComponents: GridComponents,
+      inferiorData: Map[ActorRef[
+        GridAgent.Message
+      ], (VoltageRange, Set[TransformerTapping])],
+      subnetNo: Int,
+  ): VoltageRange = {
+    // filter nodes in subnet
+    val nodesInSubnet =
+      gridComponents.nodes.filter(_.subnet == subnetNo).map(_.uuid)
+
+    // calculate voltage range
+    val nodeResMap = powerFlowResultEvent.nodeResults
+      .filter(res => nodesInSubnet.contains(res.getInputModel))
+      .map(res => res.getInputModel -> res.getvMag())
+      .toMap
+    val minVoltage = nodeResMap
+      .minByOption(_._2)
+      .getOrElse(throw new CriticalFailureException(s"No node result found!"))
+    val maxVoltage = nodeResMap
+      .maxByOption(_._2)
+      .getOrElse(throw new CriticalFailureException(s"No node result found!"))
+
+    // build initial range
+    val range = VoltageRange(
+      voltageLimits.vMax.subtract(maxVoltage._2),
+      voltageLimits.vMin.subtract(minVoltage._2),
+    )
+
+    // updating the voltage range prevent or cure line congestions
```

Comment:
```suggestion
    // updating the voltage range prevent or heal line congestions
```
cure is ok but to me heal suites better